spark-submission-worker/src/main/java/org/apache/spark/k8s/operator/SparkAppSubmissionWorker.java

```diff
@@ -0,0 +1,147 @@
+import org.apache.spark.k8s.operator.spec.ApplicationSpec;
+import org.apache.spark.k8s.operator.utils.ModelUtils;
+
+/**
+ * Similar to org.apache.spark.deploy.k8s.submit.KubernetesClientApplication. This reads args from
+ * SparkApplication instead of starting separate spark-submit process
+ */
+public class SparkAppSubmissionWorker {
+  // Default length limit for generated app id. Generated id is used as resource-prefix when
+  // user-provided id is too long for this purpose. This applied to all resources associated with
+  // the Spark app (including k8s service which has different naming length limit). This we
+  // truncate the hash part to 46 chars to leave some margin for spark resource prefix and suffix
+  // (e.g. 'spark-', '-driver-svc' . etc)
+  public static final int DEFAULT_ID_LENGTH_LIMIT = 46;
+  // Default length limit to be applied to the hash-based part of generated id
+  public static final int DEFAULT_HASH_BASED_IDENTIFIER_LENGTH_LIMIT = 36;
+  // Radix value used when generating hash-based identifier
+  public static final int DEFAULT_ENCODE_BASE = 36;
+
+  public SparkAppResourceSpec getResourceSpec(
+      SparkApplication app, KubernetesClient client, Map<String, String> confOverrides) {
+    SparkAppDriverConf appDriverConf = buildDriverConf(app, confOverrides);
+    return buildResourceSpec(appDriverConf, client);
+  }
+
+  protected SparkAppDriverConf buildDriverConf(
+      SparkApplication app, Map<String, String> confOverrides) {
+    ApplicationSpec applicationSpec = app.getSpec();
+    SparkConf effectiveSparkConf = new SparkConf();
+    if (MapUtils.isNotEmpty(applicationSpec.getSparkConf())) {
+      for (String confKey : applicationSpec.getSparkConf().keySet()) {
+        effectiveSparkConf.set(confKey, applicationSpec.getSparkConf().get(confKey));
+      }
+    }
+    if (MapUtils.isNotEmpty(confOverrides)) {
+      for (Map.Entry<String, String> entry : confOverrides.entrySet()) {
+        effectiveSparkConf.set(entry.getKey(), entry.getValue());
+      }
+    }
+    effectiveSparkConf.set("spark.kubernetes.namespace", app.getMetadata().getNamespace());
+    MainAppResource primaryResource = new JavaMainAppResource(Option.empty());
+    if (StringUtils.isNotEmpty(applicationSpec.getJars())) {
+      primaryResource = new JavaMainAppResource(Option.apply(applicationSpec.getJars()));
+      effectiveSparkConf.setIfMissing("spark.jars", applicationSpec.getJars());
+    } else if (StringUtils.isNotEmpty(applicationSpec.getPyFiles())) {
+      primaryResource = new PythonMainAppResource(applicationSpec.getPyFiles());
+      effectiveSparkConf.setIfMissing("spark.submit.pyFiles", applicationSpec.getPyFiles());
+    } else if (StringUtils.isNotEmpty(applicationSpec.getSparkRFiles())) {
+      primaryResource = new RMainAppResource(applicationSpec.getSparkRFiles());
+    }
+    effectiveSparkConf.setIfMissing(
+        "spark.master", "k8s://https://$KUBERNETES_SERVICE_HOST:$KUBERNETES_SERVICE_PORT");
```

Comment:
To @jiangzho and @aaruna, as we know, Apache Spark's `ExternalClusterManager` allows a custom K8s-based external cluster manager. So, it would be great if `Spark K8s Operator` has a capability from the beginning to allow a custom prefix in addition to `k8s` from the beginning.

https://github.com/apache/spark/blob/51623785c38c9b17a6d91cb8e7f686459bd4803e/core/src/main/scala/org/apache/spark/scheduler/ExternalClusterManager.scala#L33
```scala
def canCreate(masterURL: String): Boolean
```

Reply:
yes - it should be possible to use custom Cluster Manager by setting `spark.master`. When master is not expplicitly set in SparkConf, this would automatically generate master URL based on the environment which operator is deployed in.

To make it more handy (e.g. still support automatic generating master url but with a different prefix), a new property has been introduced. Added javaDoc in `getResourceSpec` to clarify the behavior